File: src/main/scala/utn/frba/iasc/service/AuctionService.scala
package utn.frba.iasc.service

import akka.actor.ActorRef
import akka.pattern.ask
import akka.util.Timeout
import utn.frba.iasc.actors.{CancelAuction, CreateAuction, FindAuction, FindFirstByUsername}
import utn.frba.iasc.dto.CreateAuctionDTO
import utn.frba.iasc.exception.NotFoundException
import utn.frba.iasc.model.{Auction, Buyer, Open}

import scala.concurrent.duration._
import scala.concurrent.{ExecutionContext, Future}
import scala.language.postfixOps

class AuctionService(
  private val routerActor: ActorRef,
  private implicit val executionContext: ExecutionContext
) {
  private implicit val findTimeout: Timeout = Timeout(5 seconds)

  def register(auctionDTO: CreateAuctionDTO, id: String): Future[Unit] = {
    routerActor.ask(FindFirstByUsername(auctionDTO.seller))
      .mapTo[Option[Buyer]]
      .map {
        case None => throw new NotFoundException(s"User ${auctionDTO.seller} not found")
        case Some(_) =>
          val auction = Auction(
            id = id,
            article = auctionDTO.article,
            status = Open,
            basePrice = auctionDTO.basePrice,
            seller = auctionDTO.seller
          )

          routerActor ! CreateAuction(auction, auctionDTO.maxDuration)
      }
  }

  def cancel(id: String): Unit = routerActor ! CancelAuction(id)

  def find(id: String): Future[Auction] = routerActor.ask(FindAuction(id))
    .mapTo[Option[Auction]]
    .map {
      case Some(a) => a
      case None => throw new NotFoundException(s"Auction $id not found")
    }
}

File: babylon-kt/babylon-data/src/main/kotlin/utn/frba/iasc/babylon/service/AuctionService.kt
package utn.frba.iasc.babylon.service

import utn.frba.iasc.babylon.dto.*
import utn.frba.iasc.babylon.exception.NotFoundException
import utn.frba.iasc.babylon.model.*
import utn.frba.iasc.babylon.storage.AuctionStorage
import utn.frba.iasc.babylon.storage.BuyerStorage

class AuctionService(
    private val auctionStorage: AuctionStorage,
    private val buyerStorage: BuyerStorage
) {
    fun create(createAuction: CreateAuctionDTO) {
        val auction = Auction(
            id = createAuction.auctionId,
            article = createAuction.article,
            status = Open,
            basePrice = createAuction.basePrice,
            bids = emptyList(),
            seller = createAuction.seller
        )

        auctionStorage.add(auction)
    }

    fun placeBid(placeBid: PlaceBidDTO, auctionId: String) {
        val buyer = buyerStorage.find { it.username == placeBid.buyer }
            ?: throw NotFoundException("User ${placeBid.buyer} not found")
        val auction = auctionStorage.find(auctionId)
            ?: throw NotFoundException("Auction $auctionId not found")
        val bid = Bid(
            offer = placeBid.offer,
            buyer = buyer
        )

        auctionStorage.update(auction.placeBid(bid))
    }

    fun listAuctions(): List<Auction> = auctionStorage.findAll()

    fun updateStatus(updateStatus: UpdateStatusDTO, auctionId: String) {
        val status = when (updateStatus) {
            is CancelledDTO -> Cancelled(updateStatus.cancelledOn)
            is ClosedWithWinnerDTO -> ClosedWithWinner(
                closedOn = updateStatus.closedOn,
                finalPrice = updateStatus.finalPrice,
                losers = updateStatus.losersId.map { buyerStorage.findOrThrow(it) },
                winner = buyerStorage.findOrThrow(updateStatus.winnerId)
            )
            is ClosedUnresolvedDTO -> ClosedUnresolved(updateStatus.closedOn)
        }

        val auction = auctionStorage.findOrThrow(auctionId)
        auctionStorage.update(auction.copy(status = status))
    }

    fun find(id: String): Auction = auctionStorage.find(id) ?: throw NotFoundException("Auction $id not found")
}